hash_reader.go

```diff
@@ -29,55 +29,72 @@ type HashReader interface {
 
 	// Sum appends the current hash to b and returns the resulting slice.
 	// It does not change the underlying hash state.
-	Sum(b []byte) []byte
+	Sum(b []byte) ([]byte, error)
 }
 
 type hashReader struct {
 	ionReader   ion.Reader
 	hasher      hasher
 	currentType ion.Type
+	err         error
 }
 
-func NewHashReader(ionReader ion.Reader, hasherProvider IonHasherProvider) HashReader {
-	hashReader := &hashReader{ionReader: ionReader, hasher: *newHasher(hasherProvider)}
+func NewHashReader(ionReader ion.Reader, hasherProvider IonHasherProvider) (HashReader, error) {
+	newHasher, err := newHasher(hasherProvider)
+	if err != nil {
+		return nil, err
+	}
 
-	return hashReader
+	return &hashReader{ionReader: ionReader, hasher: *newHasher}, nil
```

Comment:
We return the address because HashReader is implemented with pointer receivers. Changing the return type to (*HashReader, error) results in a build failure.

Reply:
The issue is that `*HashReader` creates a pointer **to an interface** which is not what we want. What we want is a **value** that **satisfies** the interface of `HashReader` The compiler expects a return **value** that satisfies `HashReader`. A pointer to the concrete instance of `hashReader` does that. Changing the signature to `*HashReader` sets the expectation that the return value is a pointer **to an interface** which does not make sense in Go (or at least it became an error in a version of Go, it used to be "ok" with the compiler doing extra work and dereferencing).